python/ray/tune/trial.py

```diff
@@ -45,6 +47,55 @@
 logger = logging.getLogger(__name__)
 
 
+def _serialize_dataset(dataset_input) -> Union[bytes, Dict]:
+    """Before checkpointing a trial to disk."""
+
+    def _helper(dataset_dict: dict):
+        for k, v in dataset_dict.items():
+            if isinstance(v, Dataset):
+                dataset_dict[k] = v.serialize_out_of_band()
+            elif isinstance(v, int):
+                pass
+            elif isinstance(v, dict):
+                _helper(v)
+            else:
+                raise TuneError("Unexpected dataset config before serialization!!")
+
+    _dataset = copy.deepcopy(dataset_input)
```

Comment:
Note that the out-of-band serialization of the `Dataset` will already do a deepcopy, so this deepcopy will duplicate that copy. Since `_helper` already does a full traversal of a dataset dict, you could do something like this:
```python
def _serialize_dataset(dataset_input) -> Union[bytes, Dict]:
    """Before checkpointing a trial to disk."""
    def _helper(dataset_dict: dict) -> dict:
        new = {}
        for k, v in dataset_dict.items():
            if isinstance(v, Dataset):
                new[k] = v.serialize_out_of_band()
            elif isinstance(v, int):
                new[k] = v
            elif isinstance(v, dict):
                new[k] = _helper(v)
            else:
                raise TuneError("Unexpected dataset config before serialization!!")
        return new

    if isinstance(_dataset, Dataset):
        return _dataset.serialize_out_of_band()
    elif isinstance(_dataset, dict):
        _dataset = _helper(_dataset)
        return _dataset
    else:
        raise TuneError("Unexpected dataset config before serialization!!")
```